### PR TITLE
GH-451: Add Container Stopping Error Handlers

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/451

Add error handlers that stop the container.

### CODE REVIEW

**Comment by garyrussell** [comment]
Needs a "what's new" entry after the JSON PR is merged.

**Comment by artembilan** [comment]
OK. This is much complicated than JSON one 😄 .

Too much classes anti-pattern, but I guess you have had a reason.
Or we should've never done "per-record" logic and always treat the result of the polling (`ConsumerRecords`) as a single entity...

Polling locally for review...

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingBatchErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		int n = 0;
+		while (container.isRunning() && n++ < 100) {
+			try {
+				Thread.sleep(100);
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				return;
+			}
+			throw new KafkaException("Stopped container");
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingBatchErrorHandler.java]: Something wrong is here.
You step into the loop (in most cases I believe), have some delay and ... throw exception.
You never go to the next cycle, therefore you don't wait for the container to be properly stopped.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingErrorHandler.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		while (container.isRunning() && n++ < 100) {
+			try {
+				Thread.sleep(100);
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				return;
+			}
+		}
+		throw new KafkaException("Stopped container");
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingErrorHandler.java]: OK. And here the exception is thrown outside of the loop.
This is better than previous one, but I still can't compile why should we throw exception here anyway.
And if though why doesn't it include `thrownException` then?

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerAwareErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * An error handler that has access to the unprocessed records from the last poll
+ * (including the failed record), the consumer, and the container.
+ * The records passed to the handler will not be passed to the listener
+ * (unless re-fetched if the handler performs seeks).
+ *
+ * @author Gary Russell
+ * @since 2.1
+ *
+ */
+public interface ContainerAwareErrorHandler extends RemainingRecordsErrorHandler {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerAwareErrorHandler.java]: The `@FunctionalInterface` is missing.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerAwareBatchErrorHandler.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * An error handler that has access to the batch of records from the last poll the
+ * consumer, and the container.
+ *
+ * @author Gary Russell
+ * @since 2.1
+ *
+ */
+public interface ContainerAwareBatchErrorHandler extends ConsumerAwareBatchErrorHandler {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerAwareBatchErrorHandler.java]: The @FunctionalInterface is missing.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingErrorHandler.java (hunk 2 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/ContainerStoppingErrorHandler.java]: We have to throw an exception - when an error handler "handles" an error, further processing is performed (e.g. transaction commit) but, yes, it should include the cause.

File: src/reference/asciidoc/whats-new.adoc [hunk 5]
```diff
@@ -8,3 +8,9 @@ This version requires the 1.0.0 `kafka-clients` or higher.
 
 The `StringJsonMessageConverter` and `JsonSerializer` now add type information in `Headers`, allowing the converter and `JsonDeserializer` to create specific types on reception, based on the message itself rather than a fixed configured type.
 See <<serdes>> for more information.
+
+
+==== Container Stopping Error Handlers
+
+Container Error handlers are now provided for both record and batch listeners that treat any exceptions thrown by the listener as fatal; they stop the container.
```
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: > ... and rethrow a `KafkaException` to prevent further processing.

Or something similar.

Comment (garyrussell) [reply src/reference/asciidoc/whats-new.adoc]: That's not really appropriate (too much detail) for a "what's new" entry. However, it raised an issue in my mind about the seek to current error handlers.

Polishing now...

Comment (artembilan) [reply src/reference/asciidoc/whats-new.adoc]: Oh! Really?
So, tell that in the main Doc then, please.

I think if we had it before I might not ask the question why we re-throw...

Thanks

**Comment by garyrussell** [comment]
See commit comment in last commit - parts of it must be back ported.

The interface hierarchy is required for backwards compatibility; we can consider squashing it in 3.0.

**Comment by artembilan** [comment]
> Parts of this commit will need to be back ported.

Good. Would you mind preparing an appropriate PR after merging this one?
Thanks

**Comment by garyrussell** [comment]
>Would you mind preparing an appropriate PR after merging this one?

Yes, of course; that's what I meant.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/d2ce4aba1e68de9e97a63cb0052f7c83e90e40c2 after rebasing

**Pull request closed without merging**